### ISSUE TITLE
Resuming PSK session options failing 3.0.0-beta1

### ISSUE THREAD

**Issue by tmshort** [openssl/openssl#15951]
This appears eerily similar to #7433, but for 3.0.0-beta1. Heck, the same error message is popping up! (Supposedly fixed in #7434).

Built openssl-3.0.0-beta1: `./config && make -sj`
Built OpenSSL_1_1_1-stable: `./config && make -sj`

Note that in 1.1.1-stable, the following sequence works:

I'm trying to do the same thing as #7433, create a session via PSK, and then re-use it.

Server (run once, and continues through test):
```
openssl s_client -psk 1234567890 -nocert
```

Client getting ticket:
```
openssl s_client -psk 1234567890 -sess_out mysession
```

Client using ticket:
```
openssl s_client -psk 1234567890 -sess_in mysession
```
Note: same behavior with or without the `-psk` option. I've tried various other psk options without success.

s_server error output:
```
PSK warning: client identity not what we expected (got '<garbage text>' expected 'Client_identity')                                                  
ERROR
C0113AB3757F0000:error:0A0000FD:SSL routines:tls_psk_do_binder:binder does not verify:ssl/statem/extensions.c:1603:
shutting down SSL
CONNECTION CLOSED
```
`openssl s_client -psk 1234567890 -sess_in mysession` error output:
```
CONNECTED(00000003)
C031B790BB7F0000:error:0A000417:SSL routines:ssl3_read_bytes:sslv3 alert illegal parameter:ssl/record/rec_layer_s3.c:1584:SSL alert number 47
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 7 bytes and written 660 bytes
Verification error: unspecified certificate verification error
---
New, TLSv1.3, Cipher is TLS_CHACHA20_POLY1305_SHA256
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 1 (unspecified certificate verification error)
---
```
(Connection drops)

First off, am I doing this right? It seems to work fine in 1.1.1, but not in 3.0.0-beta1.

**Comment by mattcaswell** [comment]
This issue *does* actually also impact 1.1.1 - but *only* if you additionally supply the option "-no_ticket" to the s_server command line.

The difference between the behaviour in master and 1.1.1 is due to 9c13b49a9f2, which changed `PSK_MAX_IDENTITY_LEN` from 128 to 256. It just so happens that a default OpenSSL TLSv1.3 ticket length happens to fall between those 2 values. Tickets are presented in TLSv1.3 as a PSK "identity". Passing "no_ticket" doesn't actually stop TLSv1.3 tickets completely, it just forces the use of "session ids as a ticket" instead. This significantly reduces the ticket size to below 128 in 1.1.1.

What is happening is that s_server is setting a TLSv1.2 PSK callback and a TLSv1.3 PSK callback. For backwards compat reasons the TLSv1.2 PSK callbacks also work in TLSv1.3 but are not preferred. In this scenario we use the "1234567890" PSK to create the initial connection. Subsequent to that we are doing a resumption using a TLSv1.3 ticket (psk). If the psk length is below `PSK_MAX_IDENTITY_LEN` then we first call the TLSv1.2 PSK callback. Subsequently we call the TLSv1.3 PSK callback. Unfortunately s_server's TLSv1.2 PSK callback accepts the identity regardless, even though it is an unexpected value, and hence the binder subsequently fails to verify.

**Comment by mattcaswell** [comment]
Fix in #16008

**Comment by paulidale** [comment]
The fix is merged, this should be closed.
